src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -1,14 +1,98 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     internal static class CodeRefactoringHelpers
     {
+        /// <summary>
+        /// <para>
+        /// Returns an instance of <typeparamref name="TSyntaxNode"/> for refactoring given specified selection in document or null
+        /// if no such instance exists.
+        /// </para>
+        /// <para>
+        /// A <typeparamref name="TSyntaxNode"/> instance is returned if:
+        /// - Selection is zero-width and inside/touching a Token with direct parent of type <typeparamref name="TSyntaxNode"/>.
+        /// - Selection is zero-width and touching a Token whose ancestor ends/starts precisely on current selection .
+        /// - Token whose direct parent of type <typeparamref name="TSyntaxNode"/> is selected.
+        /// - Whole node of a type <typeparamref name="TSyntaxNode"/> is selected.
+        /// </para>
+        /// <para>
+        /// Note: this function strips all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The stripped version is then used to determine relevant <see cref="SyntaxNode"/>. It also handles incomplete selections
+        /// of tokens gracefully.
+        /// </para>
+        /// </summary>
+        public static Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return TryGetSelectedNodeAsync<TSyntaxNode>(document, selection, (n) => n, cancellationToken);
+        }
+
+        public static async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, Func<SyntaxNode, SyntaxNode> extractNode, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
```

Comment:
it feels like extractNode is a heavy hammer.  Is this just to deal with a semicolon at the end of a statement?  If so, i think we literally could limit to just walking back through a semicolon as a secondary pass if we don't find anything during our first pass.

Reply:
It's not just about semicolon, the idea was that sometime selecting (way) higher node should still trigger refactorings for (way lower) node of interest. E.g. `        [|r = from i in c select i+1;|]` should IMO still trigger ConverToForeach even though the node selected is `ExpressionStatementSyntax`, AssignmentSyntax` respectively. Similar story (with very different types, however) is when it's directly part of field declaration. 

Generally that we might want to have set of helpers for extracting lower level nodes of interest out of potentially (selected) higher level nodes. AoRN the helper lives within the refactoring but that's just because I didn't know where else to put it. The idea is to have common set of extract helpers.

These helpers can't be directly inclined within the TryGetSelected… because 1) they're language specific (issue for VB/C#) and also it just doesn't feel right, tbh (we might want to extract to different depths for different refactorings (while also having clear reasonable & consistent defaults)).

Reply:
Would def be interested to get a large set of 'interesting' cases that could help drive what we think the appropriate logic would be.  THat might help out here.  Thanks!

Reply:
AoRN I can't think of any other than extracting from declarations & assignments but its usecase for those two isn't limited to ForEach. E.g. ToConstant refactoring doesn't trigger for: 

```csharp
int q = 5 + 6;[||]
[|int q = 5 + 6;|]
```

which while understandable doesn't feel intuitive (as foreach's didn't). So the extract from these two is definitely something we want to have helpers for and I can't see any clean way of doing that without the extract Func<> (can't hardcode due to tieing it to language, …)

Reply:
> which while understandable doesn't feel intuitive (as foreach's didn't). So the extract from these two is definitely something we want to have helpers for and I can't see any clean way of doing that without the extract Func<> (can't hardcode due to tieing it to language, …)

This is not the case.  We could IRefactoringHelpers be an ILanguageService, so that we could share core algorithms, but pull in language specific behavior.

Reply:
> but that one seems to (usually ?always?) have the same span as the expression

This is true for `Foo(a, b, c)` but not `Foo(ref a, ref b, out c)`

Reply:
That's a good idea.